### DISCUSSION TITLE
Accessing this.$refs.autocomplete.input to manage places api

### DISCUSSION THREAD

**Discussion by AntoineGGG** [quasarframework/quasar#10138] [General - Components / Directives / etc]
Hey at first thanks for your work, I use quasar for a few month now and really enjoy it !

I try, for google places api to access to an input with the ref "autocomplete" with this : this.$refs.autocomplete.$refs.input

And always get the same error that is dosen't exist in the DOM.
I read that we cant work with $el and $refs with vue3, so how can I figure this out ? 
Thanks for your help and if you need more information about my config feel free to ask.
Regards 
A

**Comment by ontwikkelfabriek** [comment]
@AntoineGGG 

Here are a few articles about `refs`, `el` with Vue 3 and **composition** api:
https://morioh.com/p/523b8406183e
https://learnvue.co/2021/06/the-beginners-guide-to-vue-template-refs-with-vue-3-updates/

Here are a few articles about `refs`, `el` with Vue 3 and **options** api:
https://blog.logrocket.com/vue-refs-accessing-dom-elements/

**Reply by AntoineGGG** [reply]
Thanks for the doc, unfortunatly, it still doesn't works...Still have this message "not a valid html element"...

**Comment by AntoineGGG** [comment]
To be more accurate: 

If I writte a classic input, the this.$refs.autocomplete works and then the places api works !!!
So the problem is the quasar q-input wrapper, that I can't manage to make it works....

**Comment by MilosPaunovic** [comment]
I had the same issue dealing with Autocomplete, but it was using Quasar v1.

I've solved it by creating a separate `Autocomplete` component which looked like this:

```js
<template>
  <q-card>
    <q-card-section class="q-py-none q-px-md">
      <q-input
        label="Search for a location..."
        color="dark"
        borderless
        input-class="autocomplete-input"
        data-cy="map-autocomplete"
      >
        <template v-slot:before>
          <q-icon name="room" />
        </template>
      </q-input>
    </q-card-section>
  </q-card>
</template>

<script>
export default {
  name: 'Autocomplete',
};
</script>
```

And by adding an `input-class` prop to `q-input`, in my `Map` component I've managed to initialize autocomplete functionality with this function:

```js
initializeAutocomplete(map) {
      // Get HTML element
      const input = document.getElementsByClassName('autocomplete-input')[0];

      if (input) {
        // Adding element to map and initializing autocomplete feature itself
        this.autocomplete = new google.maps.places.Autocomplete(input);

        // Bind the map's bounds (viewport) property to the autocomplete object,
        // so that the autocomplete requests use the current map bounds for the
        // bounds option in the request.
        this.autocomplete.bindTo('bounds', map);

        // Adding event listener to changing place
        this.autocomplete.addListener('place_changed', () => {
          const place = this.autocomplete.getPlace();

          if (!place.geometry) {
            // User entered the name of a Place that was not suggested and
            // pressed the Enter key, or the Place Details request failed.
            this.$q.notify({
              message: `No details available for input: '${place.name}'`,
              color: 'negative',
            });
            return;
          }

          // If the place has a geometry, then present it on a map.
          if (place.geometry.viewport) {
            map.fitBounds(place.geometry.viewport);
          } else {
            map.setCenter(place.geometry.location);
            map.setZoom(FOCUSED_ZOOM);
          }
        });
      }
    },
```

**Reply by AntoineGGG** [reply]
Thanks for sharing ! And how do you initialize this function ? in mounted ? No DOM error like "No Valid Html Element" ?

**Reply by MilosPaunovic** [reply]
Yes, I'm invoking it in mounted like so: 

```js
 async mounted() {
    const google = await mapInitialization();

    this.initializeMap(this.$el);
    this.initializeInfoWindow();
    this.initializeAutocomplete(this.map);
    
    // ...more stuff
    }
```

Because you have to wait for map itself to initialize.

To be honest, I've worked on this long ago, can't remember which exact error I had, but I can confirm it was something related to autocomplete input. And this approach solved the problem.

**Reply by AntoineGGG** [reply]
Thanks for that ! I'll try tonight and post a feedback here...I'm working on it for several hours now and want to understand where I'm wrong... have a good day !
